### PR TITLE
Verify indirect deps reachable on incremental run

### PR DESCRIPTION
**Pull request by ilevkivskyi**
Fixes https://github.com/python/mypy/issues/19477

This is another surprising correctness hole in the indirect dependencies logic. We implicitly assume everywhere that indirect dependencies are subset of transitive ones. Although it is true when the indirect dependencies are initially calculated, there is no guarantee this will stay true after an incremental update where some import structure has changed.

Currently, we only handle the situations where an indirect dependency is removed from the graph completely, but if it is still in the graph, but now brought in by a _different_ dependent, this may cause out-of-order cache loading, and thus a bug or a crash. It is interesting that although in theory it is kind of a big deal, in practice it is very hard to create such scenario. So far I found only two complicated scenarios (see tests).

Implementation note: although this additional check is needed for rare edge cases, it is quite computationally expensive. I tried few options (including only computing import deltas), but the only viable/robust option seems to be an honest recursive check with some caching. I tested this on `torch` (it has ~1000 small SCCs plus single SCC with around 1000 modules), and the performance penalty is negligible there. So hopefully it should be fine even for large code bases.

### CODE REVIEW

**Comment by ilevkivskyi** [comment]
Ping on this one.

File: mypy/build.py [hunk 1]
```diff
@@ -1201,6 +1206,19 @@ def wait_for_done_workers(self) -> tuple[list[SCC], bool, dict[str, tuple[str, l
             results,
         )
 
+    def verify_transitive(self, from_scc_id: int, to_scc_id: int) -> bool:
+        """Verify that one SCC is a (transitive) dependency of another."""
+        if (from_scc_id, to_scc_id) in self.verify_transitive_cache:
```
Comment (JukkaL) [inline mypy/build.py]: Minor: use `:=` with `.get` to avoid constructing tuple twice and performing lookup twice?

File: mypy/build.py [hunk 2]
```diff
@@ -1201,6 +1206,19 @@ def wait_for_done_workers(self) -> tuple[list[SCC], bool, dict[str, tuple[str, l
             results,
         )
 
+    def verify_transitive(self, from_scc_id: int, to_scc_id: int) -> bool:
+        """Verify that one SCC is a (transitive) dependency of another."""
```
Comment (JukkaL) [inline mypy/build.py]: Maybe a name like `is_transitive_dep` would be clearer, or `is_transitive_scc_dep`?

File: mypy/build.py [hunk 3]
```diff
@@ -1201,6 +1206,19 @@ def wait_for_done_workers(self) -> tuple[list[SCC], bool, dict[str, tuple[str, l
             results,
         )
 
+    def verify_transitive(self, from_scc_id: int, to_scc_id: int) -> bool:
```
Comment (JukkaL) [inline mypy/build.py]: Will be the cache be O(n**2) in size in the worst case? I wonder if there is a more efficient algorithm if that's the case.

Comment (ilevkivskyi) [reply mypy/build.py]: Yes. It will be quadratic in number of SCCs in worst case. I optimized for time vs space. It is a tough tradeoff. I considered few other things:
* Using SCCs processed so far: not correct because we may accidentally process unrelated SCCs, thus causing false negatives.
* Include indirect deps when ordering SCCs: not correct for roughly the same reason.
* Do not use caching: horrible time performance `O(V * E)`, i.e. `O(n**3)` worst case.
* Track imports delta, i.e. only track which dependencies changed since last build: non-trivial and error-prone, there are too many edge cases to consider.

Feel free to think about this, but I would prefer to not postpone this for long, this is a correctness issue that can cause crashes. So far we have been lucky that this didn't come often, bu there is a risk that some popular library will change their import structure hitting this issue.

In the meantime I will explore a bit more, maybe there are some niche algorithms for our specific situation, where in 99.9% there is a path in DAG between two points.

Comment (ilevkivskyi) [reply mypy/build.py]: Actually I have an idea: we can track whether SCCs are "dependency-fresh" (even if not interface-fresh), i.e. as we process them, check if all direct dependencies are the same as from cache. Essentially check if import delta is "strongly" zero, and then only fall back to full check if something changed. This will not help the worst case, but will help a lot with average case, where one simply edits some files without much changes to import structure.

Comment (ilevkivskyi) [reply mypy/build.py]: Correction: not "strongly" zero, but "strongly" positive (some extra new deps will do no harm, we just need to be sure we didn't loose any).

Comment (ilevkivskyi) [reply mypy/build.py]: @JukkaL OK, I added a performance optimization I mentioned above. Some comments:
* I only skip the expensive check when the import delta is zero. Reliably detecting import delta is positive is really tricky (if possible at all) because what constitutes the "previous run" can be different by module.
* This should actually cover a lot of cases, since if there are some changed imports it is likely the SCC will be found stale in some other way and we never get to this expensive check anyway.
* I delete an old and (IMO) dirty hack from `semanal.py` added in https://github.com/python/mypy/pull/6582. It should not be needed. I also remove the corresponding duplicate test (I verified we still have the original test and it passes).
* I add logging for the cache size in case some large code bases want to check if it is a problem.

**Comment by ilevkivskyi** [comment]
@cburroughs I added some performance optimizations to my fix, could you please check that your repros with `type-PyOpenSSL` with empty `__init__.pyi` are still fixed? (Just to be sure that I didn't break anything while optimizing.)

**Comment by cburroughs** [comment]
Still crash free; thank you!

File: mypy/build.py [hunk 4]
```diff
@@ -4201,6 +4273,21 @@ def deps_filtered(graph: Graph, vertices: AbstractSet[str], id: str, pri_max: in
...
+    """Compute stable snapshot of transitive import structure for given SCC."""
+    all_direct_deps = {
+        dep
+        for id in scc.mod_ids
+        for dep in graph[id].dependencies
+        if graph[id].priorities.get(dep) != PRI_INDIRECT
+    }
+    trans_dep_hash_map = {
+        dep_id: "" if dep_id in scc.mod_ids else graph[dep_id].trans_dep_hash.hex()
+        for dep_id in all_direct_deps
```
Comment (JukkaL) [inline mypy/build.py:4301]: Use `sorted(all_direct_deps)` to ensure consistent ordering?

File: mypy/semanal_main.py [hunk 5]
```diff
@@ -416,11 +415,6 @@ def semantic_analyze_target(
         )
         if isinstance(node, Decorator):
             infer_decorator_signature_if_simple(node, analyzer)
-    for dep in analyzer.imports:
-        state.add_dependency(dep)
-        priority = mypy.build.PRI_LOW
-        if priority <= state.priorities.get(dep, priority):
-            state.priorities[dep] = priority
```
Comment (JukkaL) [inline mypy/semanal_main.py:423]: What's this change?

File: test-data/unit/check-newsemanal.test [hunk 6]
```diff
@@ -2352,21 +2352,6 @@ class C(Generic[T]): ...
 class A: ...
 class D: ...
 
-[case testNewAnalyzerAddedSubStarImport_incremental]
```
Comment (JukkaL) [inline test-data/unit/check-newsemanal.test:2355]: What's this change?

File: mypy/build.py [hunk 7]
```diff
@@ -1203,6 +1208,20 @@ def wait_for_done_workers(self) -> tuple[list[SCC], bool, dict[str, tuple[str, l
...
+    def is_transitive_scc_dep(self, from_scc_id: int, to_scc_id: int) -> bool:
+        """Check if one SCC is a (transitive) dependency of another."""
+        edge = (from_scc_id, to_scc_id)
+        if (cached := self.transitive_deps_cache.get(edge)) is not None:
+            return cached
+        if to_scc_id in self.scc_by_id[from_scc_id].deps:
+            self.transitive_deps_cache[edge] = True
+            return True
+        for dep in self.scc_by_id[from_scc_id].deps:
+            if self.is_transitive_scc_dep(dep, to_scc_id):
```
Comment (JukkaL) [inline mypy/build.py]: This could be better without recursion, since there could be a very long import dependency chain that could blow up the stack (at least when running in interpreted mode). Probably not a big deal though.

File: mypy/build.py (hunk 4 above)
Comment (ilevkivskyi) [reply mypy/build.py:4301]: We already have `sort_keys=True` by default in `json_dumps()`, so no need to order them twice.

File: mypy/semanal_main.py (hunk 5 above)
Comment (ilevkivskyi) [reply mypy/semanal_main.py:423]: See my reply above https://github.com/python/mypy/pull/20735#discussion_r2777633899. It is a hack that is not needed anymore, it is a (very) bad idea to modify loaded graph after the initial load.

File: test-data/unit/check-newsemanal.test (hunk 6 above)
Comment (ilevkivskyi) [reply test-data/unit/check-newsemanal.test:2355]: See the same reply I mentioned above. This is a duplicate test.

File: mypy/build.py (hunk 7 above)
Comment (ilevkivskyi) [reply mypy/build.py]: Yeah, I was literally just going to push this change as we agreed yesterday.

Comment (ilevkivskyi) [reply mypy/build.py]: Btw I also deiced to use breadth-first search, it seems intuitively more efficient for common cases. Anyway, I would propose that we stop the premature optimizations here.
